### PR TITLE
Fix serialization for records which have no properties of their own

### LINKED ISSUE #8286: The serializer works incorrectly
Hey!

There is a bug in the serializer.
The explanation is a little bit complicated sorry about that :-)

**Prerequisites**
1. We have such a set of types
```csharp
  [Immutable, GenerateSerializer, Alias("TestType2")]
  public abstract record TestType2(string? TestString)
  {
      [Immutable, GenerateSerializer, Alias("TestType2A")]
      public sealed record TestType2A(string? TestString) : TestType2(TestString);
  
      [Immutable, GenerateSerializer, Alias("TestType2B")]
      public sealed record TestType2B(string? TestString) : TestType2(TestString);
  }
  
  [Immutable, GenerateSerializer, Alias("TestType1")]
  public record TestType1(TestType2[] TestArray);
  
  [Immutable, GenerateSerializer, Alias("TestType0")]
  public record TestType0(TestType1 TestType1, string? SomeString);
```
2. We have a cluster with two silos (`silo 1`, `silo 2`). Basically, we just need more than one
3. And all the data hardcoded
```csharp
public class TestSubGrain : Grain, ITestSubGrain
{
    public Task<TestType0> GetTestType0()
    {
        return Task.FromResult(
            new TestType0(TestType1:
                new TestType1(
                    TestArray: new TestType2[]
                    {
                        new TestType2.TestType2A("a"),
                        new TestType2.TestType2A("b"), // WILL BE NULL
                    }
                ),
                SomeString: "test" // WILL BE NULL
            )
        );
    }
}
```

**The problem**
When `silo 1` requests this model from `silo 2`, then `silo 1` receives nulls in some fields.

The problem is with the types `TestType2A`, `TestType2B`, and their superclass `TestType2`, and the field `TestString` they share. If we remove the field from the superclass — the problem disappears.

**There is a repro**
https://github.com/aurokk/orleans-bug-repro

### PR DESCRIPTION
**Pull request by ReubenBond**
* As the `GenerateSerializerAttribute.IncludePrimaryConstructorParameters` doc comment says, it defaults to true for `record` types - but it should only default to `true` for record types. That is fixed now. Also note that non-record types don't (yet) support primary constructors and adding one to an existing type would be a breaking change.
* If a type opts in (or not out, in the case of records) to including primary constructor parameters in serialization, then wire serialization should treat the type as potentially having primary constructor parameters even if it currently does not. Anything else would make adding a parameter a breaking change, which it is not supposed to be (version tolerance is important)

Fixes #8286 and #8275

cc @christiansparre & @ahydrax

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8289)

### CODE REVIEW

File: src/Orleans.CodeGenerator/SerializerGenerator.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Security.Cryptography.X509Certificates;
```
Comment (ahydrax) [inline src/Orleans.CodeGenerator/SerializerGenerator.cs:4]: I think this was imported accidentally

**Comment by christiansparre** [comment]
👍 I have verified the fix using our end to end tests and everything looks good.

**Comment by jkonecki** [comment]
@ReubenBond thank you for the fix! What are the plans for getting it out?

**Comment by ReubenBond** [comment]
Building and testing a release at the moment. Hopefully it goes well. I am working through some CI issues

**Comment by jkonecki** [comment]
Many thanks for the update! I'll test the fix as soon as the NuGet is
released.

**Pull request merged**